core/src/avm1/scope.rs

```diff
@@ -60,67 +59,31 @@ impl<'gc> Scope<'gc> {
     /// Construct a scope for use with `tellTarget` code where the timeline
     /// scope has been replaced with another given object.
     pub fn new_target_scope(
-        mut parent: GcCell<'gc, Self>,
+        parent: Gc<'gc, Self>,
         clip: Object<'gc>,
         mc: MutationContext<'gc, '_>,
-    ) -> GcCell<'gc, Self> {
-        let mut bottom_scope = None;
-        let mut top_scope: Option<GcCell<'gc, Self>> = None;
+    ) -> Gc<'gc, Self> {
+        let mut scope = *parent;
 
-        loop {
-            let next_scope = GcCell::allocate(
-                mc,
-                Self {
-                    parent: None,
-                    class: parent.read().class,
-                    values: parent.read().values,
-                },
-            );
-
-            if parent.read().class == ScopeClass::Target {
-                next_scope.write(mc).values = clip;
-            }
-
-            if bottom_scope.is_none() {
-                bottom_scope = Some(next_scope);
-            }
-
-            if let Some(ref scope) = top_scope {
-                scope.write(mc).parent = Some(next_scope);
-            }
-
-            top_scope = Some(next_scope);
-
-            let grandparent = parent.read().parent;
-            if let Some(grandparent) = grandparent {
-                parent = grandparent;
-            } else {
-                break;
-            }
+        if scope.class == ScopeClass::Target {
+            scope.values = clip;
+        } else {
+            scope.parent = scope.parent.map(|p| Self::new_target_scope(p, clip, mc));
```

Comment:
I think recursion can be avoided here using a simple loop.

Reply:
Unfortunately no :(
With a loop you'd need to allocate the scope chain and *then* replace the target (which is what the old code did), but now that `Scope` is immutable there's no way to replace the scope object once allocated (because it's behind a `Gc`).